### PR TITLE
[Merged by Bors] - feat: add `MultilinearMap.dfinsuppFamily` and `MultilinearMap.piFamily`

### PR DESCRIPTION
**Pull request by eric-wieser**
This is the analog to `Fintype.piFinset`

Since this is computable and rather confusing, I've included a test with a concrete evaluation.

---

- [x] depends on: #17882 (for the test)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/DFinsupp/Multilinear.lean [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+variable [DecidableEq ι] [Fintype ι] [Semiring R]
+variable [∀ i k, AddCommMonoid (M i k)] [∀ p, AddCommMonoid (N p)]
+variable [∀ i k, Module R (M i k)] [∀ p, Module R (N p)]
+
+omit [Fintype ι] in
+/-- This lemma is sufficiently impossible to name that it's easier to just keep it private. -/
+private theorem update_aux (m : (i : ι) → Π₀ (j : κ i), M i j)
+    (i : ι) (p : (i : ι) → κ i) (x : Π₀ (j : κ i), M i j) :
+    (fun k ↦ Function.update m i x k (p k)) =
+      Function.update (fun i ↦ m i (p i)) i (x (p i)) := by
```
Comment (urkud) [inline Mathlib/Data/DFinsupp/Multilinear.lean:41]: Is it a dependent version of `Function.update_apply`?

Comment (eric-wieser) [reply Mathlib/Data/DFinsupp/Multilinear.lean]: See the golf I just pushed. It's `apply_update`.

Comment (eric-wieser) [reply Mathlib/Data/DFinsupp/Multilinear.lean]: In fact, it works inline, so I've dropped this lemma.

File: Mathlib/LinearAlgebra/Multilinear/DFinsupp.lean [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    dfinsuppFamily (s • f) = s • dfinsuppFamily f := by
+  ext; simp
+
+end Semiring
+
+section CommSemiring
+
+variable [DecidableEq ι] [Fintype ι] [CommSemiring R]
+variable [∀ i k, AddCommMonoid (M i k)] [∀ i k, Module R (M i k)]
+variable [∀ p, AddCommMonoid (N p)] [∀ p, Module R (N p)]
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Multilinear/DFinsupp.lean]: Good catch, this was accidental.

File: Mathlib/LinearAlgebra/Multilinear/DFinsupp.lean [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+section Semiring
+
+variable [DecidableEq ι] [Fintype ι] [Semiring R]
+variable [∀ i k, AddCommMonoid (M i k)] [∀ p, AddCommMonoid (N p)]
+variable [∀ i k, Module R (M i k)] [∀ p, Module R (N p)]
+
+/--
+Given a family of indices `κ` and a multilinear map `f p` for each way `p` to select one index from
+each family, map a family of finitely-supported functions `x` into a finitely-supported function
+from each selection.
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Multilinear/DFinsupp.lean:40]: I have trouble parsing this sentence.

File: Mathlib/LinearAlgebra/Multilinear/Pi.lean [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+section Semiring
+
+variable [Semiring R]
+variable [∀ i k, AddCommMonoid (M i k)] [∀ p, AddCommMonoid (N p)]
+variable [∀ i k, Module R (M i k)] [∀ p, Module R (N p)]
+
+/--
+Given a family of indices `κ` and a multilinear map `f p` for each way `p` to select one index from
+each family, map a family of functions `x` into a function from each selection.
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Multilinear/Pi.lean:36]: Idem

File: Mathlib/LinearAlgebra/Multilinear/DFinsupp.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Multilinear/DFinsupp.lean:40]: Maybe

```suggestion
Given a family of indices `κ` and a multilinear map `f p` for each way `p` to select one index from
each family, `dfinsuppFamily f` maps a family of finitely-supported functions (one for each index `κ i`)
into a finitely-supported function on the selection families `Π i, κ i`.
```

**Comment by jcommelin** [comment]
bors d+

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**